src/Compilers/Core/Portable/MetadataReference/Metadata.cs

```diff
@@ -6,15 +6,34 @@
 namespace Microsoft.CodeAnalysis
 {
     /// <summary>
+    /// An Id that can be used to identify a metadata instance.  If two metadata instances 
+    /// have the same id then they are guaranteed to have the same content.  If two metadata
+    /// instances have different ids, then the contents may or may not be the same.  As such,
+    /// the id is useful as a key in a cache when a client wants to share data for a metadata
+    /// reference as long as it has not changed.
+    /// </summary>
+    public sealed class MetadataId
+    {
```

Comment:
Well, these are all made from a PortableExecutableReference.  And those have a FilePath.  So i think we could use the filepath, plus a monotonically increasing integer for that purpose.  I can make that change if you'd like that :)